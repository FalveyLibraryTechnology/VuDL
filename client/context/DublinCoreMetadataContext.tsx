import React, { createContext, useContext, useReducer } from "react";
import PropTypes from "prop-types";

export interface DublinCoreMetadata {
    currentDublinCore: Record<string, Array<string>>;
    // We need to be careful about how we set keys for elements in our control group; if we reuse
    // keys after changing components, then when we add or remove values in the list, React may
    // inappropriately reuse existing elements instead of correctly redrawing the modified list.
    // By maintaining a counter of the number of times impactful changes are made in the list, we
    // can ensure that keys are always unique, and that elements are redrawn when necessary.
    // This feels like a hack, and a better solution would be welcomed!
    keyCounter: Record<string, number>;
}

const dublinCoreContextParams: DublinCoreMetadata = { currentDublinCore: {}, keyCounter: {} };
const DublinCoreMetadataContext = createContext({});

/**
 * Update the shared states of react components.
 */
const reducerMapping: Record<string, string> = {
    SET_CURRENT_DUBLIN_CORE: "currentDublinCore",
};

const incrementKeyCounter = (keyCounter: Record<string, number>, field: string): Record<string, number> => {
    const current = keyCounter[field] ?? 0;
    return {
        ...keyCounter,
        [field]: current + 1,
    };
};

const addAbove = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index, 0, value);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const addBelow = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index + 1, 0, value);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const deleteValue = (dublinCore: Record<string, Array<string>>, field: string, index: number): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current.splice(index, 1);
    return {
        ...dublinCore,
        [field]: current,
    }
};

const replaceValue = (dublinCore: Record<string, Array<string>>, field: string, index: number, value: string): Record<string, Array<string>> => {
    const current = (dublinCore[field] ?? []).slice(0);
    current[index] = value;
    return {
        ...dublinCore,
        [field]: current,
    }
};

const dublinCoreMetadataReducer = (state: DublinCoreMetadata, { type, payload }: { type: string, payload: unknown }) => {
    if (type === "ADD_VALUE_ABOVE") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: addAbove(state.currentDublinCore, field, index, value),
        }
    } else if (type === "ADD_VALUE_BELOW") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: addBelow(state.currentDublinCore, field, index, value),
        }
    } else if (type === "DELETE_VALUE") {
        const { field, index } = payload as { field: string, index: number };
        return {
            ...state,
            keyCounter: incrementKeyCounter(state.keyCounter, field),
            currentDublinCore: deleteValue(state.currentDublinCore, field, index),
        }
    } else if (type === "REPLACE_VALUE") {
        const { field, index, value } = payload as { field: string, index: number, value: string };
        return {
            ...state,
            // We don't need to refresh the key counter in this scenario, because the number of
            // fields is not changing.
            currentDublinCore: replaceValue(state.currentDublinCore, field, index, value),
        }
    } else if(Object.keys(reducerMapping).includes(type)) {
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    }
    console.error(`processMetadata action type: ${type} does not exist`);
    return state;
};

export const DublinCoreMetadataContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(dublinCoreMetadataReducer, dublinCoreContextParams);
    const value = { state, dispatch };
    return <DublinCoreMetadataContext.Provider value={value}>{children}</DublinCoreMetadataContext.Provider>;
};

export const useDublinCoreMetadataContext = () => {
    const {
        state,
        dispatch,
    } = useContext(DublinCoreMetadataContext) as { state: DublinCoreMetadata, dispatch: (params: unknown) => void };

    const setCurrentDublinCore = (dc: Record<string, Array<string>>) => {
        dispatch({
            type: "SET_CURRENT_DUBLIN_CORE",
            payload: dc
        });
    };

    const addValueAbove = (field: string, index: number, value: string): void => {
        dispatch({
            type: "ADD_VALUE_ABOVE",
            payload: { field, index, value }
        });
    }

    const addValueBelow = (field: string, index: number, value: string): void => {
        dispatch({
            type: "ADD_VALUE_BELOW",
            payload: { field, index, value }
        });
    }

    const deleteValue = (field: string, index: number): void => {
        dispatch({
            type: "DELETE_VALUE",
            payload: { field, index }
        });
    }

    const replaceValue = (field: string, index: number, value: string): void => {
        dispatch({
            type: "REPLACE_VALUE",
            payload: { field, index, value }
        });
    }

    return {
        state,
        action: {
            addValueAbove,
            addValueBelow,
            deleteValue,
            replaceValue,
            setCurrentDublinCore,
        },
    };
};

DublinCoreMetadataContextProvider.propTypes = {
    children: PropTypes.node,
};

export default { DublinCoreMetadataContextProvider, useDublinCoreMetadataContext };
