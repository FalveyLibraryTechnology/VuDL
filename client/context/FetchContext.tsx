import React, { createContext, useContext, useReducer } from "react";
import { authApiUrl, ingestApiUrl, loginUrl } from "../util/routes";
import PropTypes from "prop-types";

/**
 * Pass a shared entity to react components,
 * specifically a way to make api requests.
 */
const fetchContextParams = {
    token: typeof window !== 'undefined' ? sessionStorage.getItem("token") : null,
};
const FetchContext = createContext({});

/**
 * Update the shared states of react components.
 */
const fetchReducer = (state, { type, payload }) => {
    switch (type) {
        case "UPDATE_TOKEN":
            sessionStorage.setItem("token", payload);
            return {
                ...state,
                token: payload,
            };
        default:
            console.error(`fetch action type: ${type} does not exist`);
            return state;
    }
};

export const FetchContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(fetchReducer, fetchContextParams);
    const value = { state, dispatch };
    return <FetchContext.Provider value={value}>{children}</FetchContext.Provider>;
};

export const useFetchContext = () => {
    const {
        state: { token },
        dispatch,
    } = useContext(FetchContext);

    /**
     * Get default fetch parameters. If params or headers exist,
     * overload the associated keys.
     */
    const fetchParams = (params = {}, headers = {}) => ({
        credentials: "include",
        method: "GET",
        mode: "cors",
        ...params,
        headers,
    });

    /**
     * Query a fresh token to authorize user to make api requests.
     */
    const refreshToken = async () => {
        try {
            const response = await fetch(`${authApiUrl}/token/mint`, fetchParams());
            if (response.ok) {
                const token = await response.json();
                dispatch({
                    type: "UPDATE_TOKEN",
                    payload: token,
                });
                return token;
            }
        } catch (error) {
            console.error(error);
        }
        window.location.href = `${loginUrl}?referer=${encodeURIComponent(window.location.href)}`;
        return false;
    };

    /**
     * Return a customizable response when making a request.
     * If first response return a 401 unauthorized status,
     * attempt to refresh token and call the request again.
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const makeRequest = async (url, params = {}, headers = {}) => {
        const response = await fetch(
            url,
            fetchParams(params, {
                ...headers,
                Authorization: `Token ${token}`,
            })
        );
        if (response.status == 401) {
            const token = await refreshToken();
            return token
                ? await fetch(
                      url,
                      fetchParams(params, {
                          ...headers,
                          Authorization: `Token ${token}`,
                      })
                  )
                : false;
        }
        return response;
    };

    /**
     * Return a json response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchJSON = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response?.ok) {
            return await response.json();
        }
        throw new Error(response?.statusText ?? "Could not make request");
    };

    /**
     * Return a text response when making a request
     * @param {string} url - The http url of the request
     * @param {Object} params - The request parameters
     * @param {Object} headers - The request parameter headers
     */
    const fetchText = async (url = ingestApiUrl, params = {}, headers = {}) => {
        const response = await makeRequest(url, params, headers);
        if (response.ok) {
            return await response.text();
        }
        throw new Error(response.statusText);
    };

    return {
        state: {
            token,
        },
        action: {
            fetchJSON,
            fetchText,
            makeRequest,
        },
    };
};

FetchContextProvider.propTypes = {
    children: PropTypes.node,
};

export default { FetchContextProvider, useFetchContext };
