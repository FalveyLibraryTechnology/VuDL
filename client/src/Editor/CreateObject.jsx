import React from "react";
import { TreeView, TreeItem } from "@material-ui/lab";

import AjaxHelper from "../AjaxHelper";

export default class CreateObject extends React.Component {
    constructor(props) {
        super(props);
        this.ajax = AjaxHelper.getInstance();
        this.state = { models: {}, selectedModel: null };
        // BIND
        this.handleSelect = this.handleSelect.bind(this);
    }

    componentDidMount() {
        this.ajax.getJSONPromise(this.ajax.apiUrl + "/edit/models").then((json) => {
            this.setState({ models: json });
        });
    }

    renderTree(nodes) {
        return (
            <TreeItem key={nodes.id} nodeId={nodes.id} label={nodes.name}>
                {Array.isArray(nodes.children) ? nodes.children.map((node) => this.renderTree(node)) : null}
            </TreeItem>
        );
    }

    handleSelect(event, model) {
        event.preventDefault();
        // Ignore categories
        if (model.slice(0, 2) == "__") {
            return;
        }
        this.setState({ selectedModel: model });
        return false;
    }

    render() {
        let categories = [];
        for (let category in this.state.models) {
            let children = [];
            for (let model in this.state.models[category]) {
                let value = this.state.models[category][model];
                children.push(<TreeItem key={model} nodeId={value} label={model} />);
            }
            categories.push(
                <TreeItem key={category} nodeId={`__${category}`} label={category}>
                    {children}
                </TreeItem>
            );
        }
        return (
            <form className="editor__create-object" action={this.ajax.apiUrl + "/object/new"}>
                <label>
                    Title
                    <input type="text" name="title" required />
                </label>

                <label>
                    Parent ID
                    <input type="text" name="parent" required />
                </label>

                <label>
                    <input type="radio" name="active" value="1" />
                    Active
                </label>

                <label>
                    <input type="radio" name="active" value="0" checked />
                    Inactive
                </label>

                <label>
                    Select Model Type:
                    <b>{this.state.selectedModel}</b>
                </label>
                <TreeView defaultCollapseIcon={"➖"} defaultExpandIcon={"➕"} onNodeSelect={this.handleSelect}>
                    {categories}
                </TreeView>

                {this.state.selectedModel && <button>Create Object</button>}
            </form>
        );
    }
}
