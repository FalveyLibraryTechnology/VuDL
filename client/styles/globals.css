:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
