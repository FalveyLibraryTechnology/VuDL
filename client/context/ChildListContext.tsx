import React, { createContext, useContext, useReducer } from "react";
import { getObjectChildrenUrl } from "../util/routes";
import { useFetchContext } from "./FetchContext";

interface Children {
    numFound?: number;
    start?: number;
    docs?: Record<string, string>[];
}

interface ChildListState {
    childStorage: Record<string, Children>;
}

const childListContextParams: ChildListState = {
    childStorage: {},
};

const ChildListContext = createContext({});

const reducerMapping: Record<string, string> = {
};

/**
 * Update the shared states of react components.
 */
const childListReducer = (state: ChildListState, { type, payload }: { type: string, payload: unknown}) => {
    if (type === "ADD_TO_CHILD_STORAGE") {
        const { key, children } = payload;
        const childStorage = {
            ...state.childStorage,
        };
        childStorage[key] = children;
        return {
            ...state,
            childStorage
        };
    } else if (Object.keys(reducerMapping).includes(type)){
        return {
            ...state,
            [reducerMapping[type]]: payload
        };
    } else {
        console.error(`fetch action type: ${type} does not exist`);
        return state;
    }
};

export const ChildListContextProvider = ({ children }) => {
    const [state, dispatch] = useReducer(childListReducer, childListContextParams);
    const value = { state, dispatch };
    return <ChildListContext.Provider value={value}>{children}</ChildListContext.Provider>;
};

export const useChildListContext = () => {
    const {
        action: {
            fetchJSON
        }
    }= useFetchContext();
    const {
        state: {
            childStorage
        },
        dispatch,
    } = useContext(ChildListContext);

    const addToChildStorage = (key: string, children: Children) => {
        dispatch({
            type: "ADD_TO_CHILD_STORAGE",
            payload: { key, children },
        });
    };

    const getChildStorageKey = (pid: string, page: number, pageSize: number): string => {
        return `${pid}_${page}_${pageSize}`;
    }

    const loadChildrenIntoStorage = async (pid: string, page: number, pageSize: number) => {
        const key = getChildStorageKey(pid, page, pageSize);
        const url = getObjectChildrenUrl(pid, (page - 1) * pageSize, pageSize);
        try {
            addToChildStorage(key, await fetchJSON(url));
        } catch (e) {
            console.error("Problem fetching tree data from " + url);
        }
    }

    return {
        state: {
            childStorage,
        },
        action: {
            getChildStorageKey,
            loadChildrenIntoStorage,
        },
    };
}

export default {
    ChildListContextProvider,
    useChildListContext
}