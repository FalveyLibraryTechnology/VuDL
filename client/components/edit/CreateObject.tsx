import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { TreeView, TreeItem } from "@material-ui/lab";
import { useFetchContext } from "../../context/FetchContext";
import { apiUrl } from "../../util/routes";
import Link from "next/link";

const CreateObject = ({ parentPid = "", allowNoParentPid = false, allowChangeParentPid = true }) => {
    // Validate properties
    if (!allowChangeParentPid && !allowNoParentPid && parentPid === "") {
        throw new Error("allowChangeParentPid and allowNoParentPid cannot both be false when parentPid is empty.");
    }
    if (!allowChangeParentPid && allowNoParentPid && parentPid !== "") {
        throw new Error("allowNoParentPid=true requires allowChangeParentPid to be true when parentPid is non-empty.");
    }
    const {
        action: { makeRequest, fetchJSON },
    } = useFetchContext();
    const [models, setModels] = useState({});
    const [selectedModel, setSelectedModel] = useState("");
    const [results, setResults] = useState();
    const [title, setTitle] = useState("");
    const [parent, setParent] = useState(parentPid);
    const [noParent, setNoParent] = useState(allowNoParentPid && parentPid === "");
    const [state, setState] = useState("Inactive");

    const states = ["Active", "Inactive", "Deleted"];

    useEffect(() => {
        async function loadModels() {
            let models = {};
            try {
                models = await fetchJSON(apiUrl + "/edit/models");
            } catch (e) {
                console.error("Problem fetching models: ", e.message);
            }
            setModels(models);
        }
        loadModels();
    }, []);

    function handleSelect(event, model) {
        event.preventDefault();
        // Ignore categories
        if (model.slice(0, 2) == "__") {
            return;
        }
        setSelectedModel(model);
        return false;
    }

    async function handleSubmit(event) {
        event.preventDefault();
        const url = apiUrl + "/edit/object/new";
        const data = {
            title,
            parent,
            model: selectedModel,
            state,
        };
        setResults("Working...");
        try {
            const params = { method: "POST", body: JSON.stringify(data) };
            const headers = { "Content-Type": "application/json" };
            const result = await makeRequest(url, params, headers);
            const pid = await result.text();
            setResults(
                <span>
                    Object created: <Link href={"/edit/object/" + pid}>{pid}</Link>
                </span>
            );
        } catch (e) {
            setResults("Error: " + e.message);
        }
    }

    function handleNoParentChange(event) {
        setNoParent(event.target.checked);
        if (allowChangeParentPid && event.target.checked) {
            setParent("");
        }
    }

    function handleParentChange(event) {
        setParent(event.target.value);
        if (allowNoParentPid) {
            setNoParent(event.target.value.length === 0);
        }
    }

    function handleStateChange(event) {
        setState(event.target.value);
    }

    function handleTitleChange(event) {
        setTitle(event.target.value);
    }

    let categories = [];
    for (let category in models) {
        let children = [];
        for (let model in models[category]) {
            let value = models[category][model];
            children.push(<TreeItem key={model} nodeId={value} label={model} />);
        }
        categories.push(
            <TreeItem key={category} nodeId={`__${category}`} label={category}>
                {children}
            </TreeItem>
        );
    }
    const stateControls = states.map((currentState) => (
        <label key={currentState + "-label"}>
            {currentState}
            <input
                type="radio"
                name="state"
                value={currentState}
                checked={state === currentState}
                onChange={handleStateChange}
            />
        </label>
    ));
    let noParentControl = "";
    if (allowNoParentPid) {
        noParentControl = allowChangeParentPid ? (
            <label>
                <input type="checkbox" name="noParent" value="1" onChange={handleNoParentChange} checked={noParent} />
                No parent PID
            </label>
        ) : (
            <input type="hidden" name="noParent" value="1" />
        );
    }
    let parentControl = "";
    if (allowChangeParentPid || parent.length > 0) {
        parentControl = allowChangeParentPid ? (
            <label>
                Parent ID
                <input type="text" value={parent} name="parent" onChange={handleParentChange} required={!noParent} />
            </label>
        ) : (
            <input type="hidden" value={parent} name="parent" />
        );
    }
    // If the form has completed, just display the result:
    if (results) {
        return results;
    }
    return (
        <form onSubmit={handleSubmit} className="editor__create-object">
            <label>
                Title
                <input type="text" value={title} name="title" onChange={handleTitleChange} required />
            </label>

            {noParentControl}
            {parentControl}
            {stateControls}

            <label>
                Select Model Type:
                <b>{selectedModel}</b>
            </label>
            <TreeView defaultCollapseIcon={"➖"} defaultExpandIcon={"➕"} onNodeSelect={handleSelect}>
                {categories}
            </TreeView>

            {selectedModel && <button>Create Object</button>}
        </form>
    );
};

CreateObject.propTypes = {
    parentPid: PropTypes.string,
    allowChangeParentPid: PropTypes.bool,
    allowNoParentPid: PropTypes.bool,
};

export default CreateObject;
