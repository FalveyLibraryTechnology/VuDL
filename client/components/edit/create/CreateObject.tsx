import React, { useEffect, useState } from "react";
import { useFetchContext } from "../../../context/FetchContext";
import { apiUrl, newEditObjectUrl } from "../../../util/routes";
import Link from "next/link";
import ParentControl from "../../shared/ParentControl";
import NoParentControl from "../../shared/NoParentControl";
import StateControls from "../../shared/StateControls";
import CategoryTreeView from "./CategoryTreeView";

interface CreateObjectProps {
    parentPid: string;
    allowChangeParentPid: boolean;
    allowNoParentPid: boolean;
}

const CreateObject = ({
    parentPid = "",
    allowNoParentPid = false,
    allowChangeParentPid = true,
}: CreateObjectProps): React.ReactElement => {
    const {
        action: { fetchText, fetchJSON },
    } = useFetchContext();
    const [models, setModels] = useState({});
    const [selectedModel, setSelectedModel] = useState("");
    const [results, setResults] = useState({
        state: "initial",
        content: "",
    });
    const [title, setTitle] = useState("");
    const [currentState, setCurrentState] = useState("Inactive");
    const [noParent, setNoParent] = useState(allowNoParentPid && parentPid === "");
    const [parent, setParent] = useState(parentPid);
    const handleParentChange = (event) => {
        setParent(event.target.value);
        if (allowNoParentPid) {
            setNoParent(event.target.value.length === 0);
        }
    };

    const handleNoParentChange = (event) => {
        setNoParent(event.target.checked);
        if (allowChangeParentPid && event.target.checked) {
            setParent("");
        }
    };

    useEffect(() => {
        async function loadModels() {
            let models = {};
            try {
                models = await fetchJSON(apiUrl + "/edit/models");
            } catch (e) {
                const error = e as Error;
                console.error("Problem fetching models: ", error.message);
            }
            setModels(models);
        }
        loadModels();
    }, []);

    const handleSubmit = async (event) => {
        event.preventDefault();
        setResults({
            state: "loading",
            content: "Working...",
        });
        try {
            const params = {
                method: "POST",
                body: JSON.stringify({
                    title,
                    parent,
                    model: selectedModel,
                    state: currentState,
                }),
            };
            const headers = { "Content-Type": "application/json" };
            const pid = await fetchText(newEditObjectUrl, params, headers);
            setResults({
                state: "success",
                content: pid,
            });
        } catch (e) {
            setResults({
                state: "error",
                content: "Error: " + e.message,
            });
        }
    };

    switch (results.state) {
        case "loading":
        case "error":
            return <>{results.content}</>;
        case "success":
            return (
                <span>
                    Object created: <Link href={"/edit/object/" + results.content}>{results.content}</Link>
                </span>
            );
        default:
            return (
                <form onSubmit={handleSubmit} className="editor__create-object">
                    <label>
                        Title
                        <input
                            type="text"
                            value={title}
                            name="title"
                            onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
                                setTitle(event.target.value);
                            }}
                            required
                        />
                    </label>

                    {allowNoParentPid && allowChangeParentPid && (
                        <NoParentControl handleNoParentChange={handleNoParentChange} noParent={noParent} />
                    )}

                    {allowChangeParentPid && (
                        <ParentControl parent={parent} handleParentChange={handleParentChange} noParent={noParent} />
                    )}
                    <StateControls currentState={currentState} setCurrentState={setCurrentState} />

                    <label>
                        Select Model Type:
                        <b>{selectedModel}</b>
                    </label>
                    <CategoryTreeView models={models} setSelectedModel={setSelectedModel} />
                    {selectedModel && <button>Create Object</button>}
                </form>
            );
    }
};

export default CreateObject;
