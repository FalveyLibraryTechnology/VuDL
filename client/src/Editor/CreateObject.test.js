import React from "react";
import { beforeEach, describe, expect, it, jest } from "@jest/globals";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import { mount, render } from "enzyme";
import toJson from "enzyme-to-json";
import CreateObject from "./CreateObject";
import { FetchContextProvider, useFetchContext } from "../context";

let nodeSelectFunction = null;
let treeItems = null;
jest.mock("@material-ui/lab", function () {
    return {
        TreeView: ({ onNodeSelect, children }) => {
            nodeSelectFunction = onNodeSelect;
            treeItems = children;
            return children;
        },
        TreeItem: () => "TreeItem",
    };
});

function setFakeModels(callback) {
    callback(["model-foo", "model-bar", "model-baz"]);
}

describe("CreateObject", () => {
    let ajax;
    let props;

    beforeEach(() => {
        props = {
            parentPid: "",
            allowNoParentPid: false,
            allowChangeParentPid: true,
        };
        ajax = {};
        useFetchContext.fetchJSON = jest.fn(() => new Promise(setFakeModels, jest.fn()));
        useFetchContext.makeRequest = jest.fn();
    });

    function getCreateObjectToTest(props) {
        return <FetchContextProvider><CreateObject {...props} /></FetchContextProvider>;
    }

    it("renders appropriately with default settings", async () => {
        const wrapper = render(getCreateObjectToTest(props));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders appropriately with noParent and parent change enabled", () => {
        props.allowNoParentPid = true;
        const wrapper = render(getCreateObjectToTest(props));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders appropriately with editing disabled", () => {
        props.parentPid = "foo:1234";
        props.allowChangeParentPid = false;
        const wrapper = render(getCreateObjectToTest(props));
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("disallows all empty pid settings", () => {
        props.allowChangeParentPid = false;
        expect(() => render(getCreateObjectToTest(props))).toThrowError(
            "allowChangeParentPid and allowNoParentPid cannot both be false when parentPid is empty."
        );
    });

    it("disallows incompatible allowNoParentPid/allowChangeParentPid settings", () => {
        props.allowNoParentPid = true;
        props.allowChangeParentPid = false;
        props.parentPid = "foo:pid";
        expect(() => render(getCreateObjectToTest(props))).toThrowError(
            "allowNoParentPid=true requires allowChangeParentPid to be true when parentPid is non-empty."
        );
    });

    it("submits appropriate data in default case", async () => {
        let wrapper;
        await waitFor(() => {
            wrapper = mount(getCreateObjectToTest(props));
        });
        act(() => {
            nodeSelectFunction(new Event("event-foo"), "model-foo");
            // Test that category select has no effect:
            nodeSelectFunction(new Event("event-foo"), "__categoryWillBeIgnored");
        });
        wrapper.find("input[name='title']").simulate("change", { target: { value: "Test Title" } });
        wrapper.find("input[name='parent']").simulate("change", { target: { value: "foo:1234" } });
        wrapper.find("form").simulate("submit");
        expect(treeItems.length).toEqual(3); // make sure setFakeModels is working
        expect(useFetchContext.makeRequest).toHaveBeenCalledWith(
            "http://foo/edit/object/new",
            expect.objectContaining({
                method: "POST",
                data: JSON.stringify({
                    model: "model-foo",
                    parent: "foo:1234",
                    state: "Inactive",
                    title: "Test Title",
                }),
            }),
            { "Content-Type": "application/json" }
        );
    });

    it("pre-fills parent pid using parentPid property", async () => {
        props.parentPid = "foo:1234";
        props.allowChangeParentPid = false;
        let wrapper;
        await waitFor(() => {
            wrapper = mount(getCreateObjectToTest(props));
        });
        act(() => {
            nodeSelectFunction(new Event("event-foo"), "model-foo");
        });
        wrapper.find("input[name='title']").simulate("change", { target: { value: "Test Title" } });
        wrapper.find("form").simulate("submit");
        expect(treeItems.length).toEqual(3); // make sure setFakeModels is working
        expect(useFetchContext.makeRequest).toHaveBeenCalledWith(
            "http://foo/edit/object/new",
            expect.objectContaining({
                method: "POST",
                data: JSON.stringify({
                    model: "model-foo",
                    parent: "foo:1234",
                    state: "Inactive",
                    title: "Test Title",
                }),
            }),
            { "Content-Type": "application/json" }
        );
    });

    it("submits appropriate data with active state and no parent", async () => {
        props.parentPid = "foo:1234";
        props.allowNoParentPid = true;
        let wrapper;
        await waitFor(() => {
            wrapper = mount(getCreateObjectToTest(props));
        });
        act(() => {
            nodeSelectFunction(new Event("event-foo"), "model-foo");
        });
        wrapper.find("input[name='title']").simulate("change", { target: { value: "Test Title" } });
        wrapper.find("input[name='state'][value='Active']").simulate("change", { target: { value: "Active" } });
        wrapper.find("input[name='noParent']").simulate("change", { target: { checked: true } });
        wrapper.find("form").simulate("submit");
        expect(useFetchContext.makeRequest).toHaveBeenCalledWith(
            "http://foo/edit/object/new",
            expect.objectContaining({
                method: "POST",
                data: JSON.stringify({
                    model: "model-foo",
                    parent: "",
                    state: "Active",
                    title: "Test Title",
                }),
            }),
            { "Content-Type": "application/json" }
        );
    });

    it("checks no parent when parent pid is cleared", async () => {
        props.parentPid = "foo:1234";
        props.allowNoParentPid = true;
        let wrapper;
        await waitFor(() => {
            wrapper = mount(getCreateObjectToTest(props));
        });
        act(() => {
            nodeSelectFunction(new Event("event-foo"), "model-foo");
        });
        wrapper.find("input[name='title']").simulate("change", { target: { value: "Test Title" } });
        wrapper.find("input[name='state'][value='Active']").simulate("change", { target: { value: "Active" } });
        wrapper.find("input[name='parent']").simulate("change", { target: { value: "" } });
        wrapper.find("form").simulate("submit");
        expect(useFetchContext.makeRequest).toHaveBeenCalledWith(
            "http://foo/edit/object/new",
            expect.objectContaining({
                method: "POST",
                data: JSON.stringify({
                    model: "model-foo",
                    parent: "",
                    state: "Active",
                    title: "Test Title",
                }),
            }),
            { "Content-Type": "application/json" }
        );
    });
});
