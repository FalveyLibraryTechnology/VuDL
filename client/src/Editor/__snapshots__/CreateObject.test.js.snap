// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateObject renders appropriately with default settings 1`] = `
<form
  class="editor__create-object"
>
  <label>
    Title
    <input
      name="title"
      required=""
      type="text"
      value=""
    />
  </label>
  <label>
    Parent ID
    <input
      name="parent"
      required=""
      type="text"
      value=""
    />
  </label>
  <label>
    Active
    <input
      name="state"
      type="radio"
      value="Active"
    />
  </label>
  <label>
    Inactive
    <input
      checked=""
      name="state"
      type="radio"
      value="Inactive"
    />
  </label>
  <label>
    Deleted
    <input
      name="state"
      type="radio"
      value="Deleted"
    />
  </label>
  <label>
    Select Model Type:
    <b />
  </label>
</form>
`;

exports[`CreateObject renders appropriately with editing disabled 1`] = `
<form
  class="editor__create-object"
>
  <label>
    Title
    <input
      name="title"
      required=""
      type="text"
      value=""
    />
  </label>
  <input
    name="parent"
    type="hidden"
    value="foo:1234"
  />
  <label>
    Active
    <input
      name="state"
      type="radio"
      value="Active"
    />
  </label>
  <label>
    Inactive
    <input
      checked=""
      name="state"
      type="radio"
      value="Inactive"
    />
  </label>
  <label>
    Deleted
    <input
      name="state"
      type="radio"
      value="Deleted"
    />
  </label>
  <label>
    Select Model Type:
    <b />
  </label>
</form>
`;

exports[`CreateObject renders appropriately with noParent and parent change enabled 1`] = `
<form
  class="editor__create-object"
>
  <label>
    Title
    <input
      name="title"
      required=""
      type="text"
      value=""
    />
  </label>
  <label>
    <input
      checked=""
      name="noParent"
      type="checkbox"
      value="1"
    />
    No parent PID
  </label>
  <label>
    Parent ID
    <input
      name="parent"
      type="text"
      value=""
    />
  </label>
  <label>
    Active
    <input
      name="state"
      type="radio"
      value="Active"
    />
  </label>
  <label>
    Inactive
    <input
      checked=""
      name="state"
      type="radio"
      value="Inactive"
    />
  </label>
  <label>
    Deleted
    <input
      name="state"
      type="radio"
      value="Deleted"
    />
  </label>
  <label>
    Select Model Type:
    <b />
  </label>
</form>
`;
